# Capstone 1: Data Wrangling (from CSV)

### Data Acquisition: Get source data from CSV file

In [1]:
# Import packages and modules
import pandas as pd
import numpy as np
import datetime

In [2]:
# Assign file path
"""
    The original CSV file can be exported from 
    https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj
"""
file_path = r''

# Read CSV data into a Pandas dataframe
df = pd.read_csv(file_path,header=0)

df.columns # list all column header names

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (14,15,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Index(['CAD_INCIDENT_ID', 'INCIDENT_DATETIME', 'INITIAL_CALL_TYPE',
       'INITIAL_SEVERITY_LEVEL_CODE', 'FINAL_CALL_TYPE',
       'FINAL_SEVERITY_LEVEL_CODE', 'FIRST_ASSIGNMENT_DATETIME',
       'VALID_DISPATCH_RSPNS_TIME_INDC', 'DISPATCH_RESPONSE_SECONDS_QY',
       'FIRST_ACTIVATION_DATETIME', 'FIRST_ON_SCENE_DATETIME',
       'VALID_INCIDENT_RSPNS_TIME_INDC', 'INCIDENT_RESPONSE_SECONDS_QY',
       'INCIDENT_TRAVEL_TM_SECONDS_QY', 'FIRST_TO_HOSP_DATETIME',
       'FIRST_HOSP_ARRIVAL_DATETIME', 'INCIDENT_CLOSE_DATETIME',
       'HELD_INDICATOR', 'INCIDENT_DISPOSITION_CODE', 'BOROUGH', 'ATOM',
       'INCIDENT_DISPATCH_AREA', 'ZIPCODE', 'POLICEPRECINCT',
       'CITYCOUNCILDISTRICT', 'COMMUNITYDISTRICT', 'COMMUNITYSCHOOLDISTRICT',
       'CONGRESSIONALDISTRICT', 'REOPEN_INDICATOR', 'SPECIAL_EVENT_INDICATOR',
       'STANDBY_INDICATOR', 'TRANSFER_INDICATOR'],
      dtype='object')

In [3]:
# Convert all column header names to lowercase
col_names = [str(name).lower() for name in list(df.columns)]
df.columns = col_names
df.columns

Index(['cad_incident_id', 'incident_datetime', 'initial_call_type',
       'initial_severity_level_code', 'final_call_type',
       'final_severity_level_code', 'first_assignment_datetime',
       'valid_dispatch_rspns_time_indc', 'dispatch_response_seconds_qy',
       'first_activation_datetime', 'first_on_scene_datetime',
       'valid_incident_rspns_time_indc', 'incident_response_seconds_qy',
       'incident_travel_tm_seconds_qy', 'first_to_hosp_datetime',
       'first_hosp_arrival_datetime', 'incident_close_datetime',
       'held_indicator', 'incident_disposition_code', 'borough', 'atom',
       'incident_dispatch_area', 'zipcode', 'policeprecinct',
       'citycouncildistrict', 'communitydistrict', 'communityschooldistrict',
       'congressionaldistrict', 'reopen_indicator', 'special_event_indicator',
       'standby_indicator', 'transfer_indicator'],
      dtype='object')

A downloadable description of each dataset field is available at https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj in the _Attachments_ section under the file name **EMS_incident_dispatch_data_description.xlsx**. 

### Inspect dataframe

In [4]:
df.shape

(8557848, 32)

In [5]:
df.head()

cad_incident_id       incident_datetime initial_call_type  \
0        130010001  01/01/2013 12:00:04 AM            RESPIR   
1        130010002  01/01/2013 12:00:19 AM              CARD   
2        130010004  01/01/2013 12:01:04 AM            ARREST   
3        130010005  01/01/2013 12:01:16 AM              SICK   
4        130010006  01/01/2013 12:01:26 AM            INJURY   

   initial_severity_level_code final_call_type  final_severity_level_code  \
0                            4          RESPIR                          4   
1                            3            CARD                          3   
2                            1          ARREST                          1   
3                            6            SICK                          6   
4                            5          INJURY                          5   

  first_assignment_datetime valid_dispatch_rspns_time_indc  \
0    01/01/2013 12:01:45 AM                              Y   
1    01/01/2013 12:01:18 AM                              Y   
2    01/01/2013 12:01:33 AM                              Y   
3    01/01/2013 12:02:12 AM                              Y   
4    01/01/2013 12:01:58 AM                              Y   

   dispatch_response_seconds_qy first_activation_datetime        ...          \
0                           101    01/01/2013 12:01:51 AM        ...           
1                            59    01/01/2013 12:02:08 AM        ...           
2                            29    01/01/2013 12:01:58 AM        ...           
3                            56    01/01/2013 12:02:55 AM        ...           
4                            32    01/01/2013 12:02:55 AM        ...           

   zipcode policeprecinct  citycouncildistrict  communitydistrict  \
0  10472.0           43.0                 18.0              209.0   
1  10454.0           40.0                  8.0              201.0   
2  11418.0          102.0                 29.0              409.0   
3  10453.0           46.0                 14.0              205.0   
4  10457.0           48.0                 15.0              206.0   

  communityschooldistrict congressionaldistrict reopen_indicator  \
0                    12.0                  15.0                N   
1                     7.0                  15.0                N   
2                    27.0                   5.0                N   
3                    10.0                  15.0                N   
4                    10.0                  15.0                N   

  special_event_indicator  standby_indicator transfer_indicator  
0                       N                  N                  N  
1                       N                  N                  N  
2                       N                  N                  N  
3                       N                  N                  N  
4                       N                  N                  N  

[5 rows x 32 columns]

In [6]:
df.info(verbose=True,null_counts=True,memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8557848 entries, 0 to 8557847
Data columns (total 32 columns):
cad_incident_id                   8557848 non-null int64
incident_datetime                 8557848 non-null object
initial_call_type                 8557848 non-null object
initial_severity_level_code       8557848 non-null int64
final_call_type                   8557848 non-null object
final_severity_level_code         8557848 non-null int64
first_assignment_datetime         8498783 non-null object
valid_dispatch_rspns_time_indc    8557848 non-null object
dispatch_response_seconds_qy      8557848 non-null int64
first_activation_datetime         8483588 non-null object
first_on_scene_datetime           8284662 non-null object
valid_incident_rspns_time_indc    8557848 non-null object
incident_response_seconds_qy      8283276 non-null float64
incident_travel_tm_seconds_qy     8284535 non-null float64
first_to_hosp_datetime            6031607 non-null object
first_hosp_arrival_

In [7]:
df.memory_usage(deep=True)

Index                                    80
cad_incident_id                    68462784
incident_datetime                 676069992
initial_call_type                 529894033
initial_severity_level_code        68462784
final_call_type                   529993248
final_severity_level_code          68462784
first_assignment_datetime         673293937
valid_dispatch_rspns_time_indc    530586576
dispatch_response_seconds_qy       68462784
first_activation_datetime         672579772
first_on_scene_datetime           663230250
valid_incident_rspns_time_indc    530586576
incident_response_seconds_qy       68462784
incident_travel_tm_seconds_qy      68462784
first_to_hosp_datetime            557336665
first_hosp_arrival_datetime       555408114
incident_close_datetime           675918323
held_indicator                    530586576
incident_disposition_code          68462784
borough                           554812429
atom                              478931352
incident_dispatch_area          

### Data Cleansing: Define functions

REMOVE UNNECESSARY VALUES (**drop_errors** and **drop_immaterial**)

For the purpose of this analysis, any observation with a missing value for 'incident_disposition_code' must be omitted since the target variable is derived from this feature. In addition, observations that contain the following outliers, errors, or immaterial information must also be removed from the dataset:
+ incidents created to transport a patient from one facility to another
+ incidents where units were assigned to stand by in case they were needed
+ incidents that pertain to special events
+ incidents that were once closed but later reopened
+ incidents with calculation errors for duration metrics
+ features that contain redundant geographic information for incident

In [8]:
def drop_errors(dfObj):
    """
    This function drops rows and columns within the DataFrame 
    object that will confound future analyses
    
    Parameter(s)
    ------------
    dfObj: Pandas DataFrame object
        The object to be modified by the function
        
    Returns
    -------
    dfObj: Pandas DataFrame object
    """
    # Drop all rows with missing value for 'incident_disposition_code'
    dfObj.dropna(subset=['incident_disposition_code'],inplace=True)
    
    # Identify all columns that validate duration metrics
    list_of_validation_cols = [name for name in list(dfObj.columns)
                               if 'valid' in str(name)]
    
    # Drop all rows with invalid duration metrics
    for name in list_of_validation_cols:
        invalid_idx = dfObj[dfObj[name]=='N'].index
        dfObj.drop(invalid_idx, inplace=True)
    
    # Drop all rows where EMS were not dispatched
    no_disp_idx = dfObj[dfObj.dispatch_response_seconds_qy==0].index
    dfObj.drop(no_disp_idx, inplace=True)
    
    return dfObj

In [9]:
def drop_immaterial(dfObj):
    """
    This function drops rows and columns within the DataFrame 
    object that relates to erroneous information
    
    Parameter(s)
    ------------
    dfObj: Pandas DataFrame object
        The object to be modified by the function
        
    Returns
    -------
    dfObj: Pandas DataFrame object
    """
    # Identify all columns with outlier event indicators
    list_of_indicator_cols = [name for name in list(dfObj.columns) 
                              if 'indicator' in str(name) and name !='held_indicator']
    
    # Drop all rows that pertain to outlier incidents
    for name in list_of_indicator_cols:
        outlier_idx = dfObj[dfObj[name]=='Y'].index
        dfObj.drop(outlier_idx, inplace=True)
    
    # Remove columns that contain incident indicator data
    dfObj.drop(list_of_indicator_cols,axis=1,inplace=True)
    
    # Identify and remove all columns that contain redundant geographic data
    list_of_zone_cols = [name for name in list(dfObj.columns) 
                         if ('district' in str(name) or name=='policeprecinct')]
    dfObj.drop(list_of_zone_cols,axis=1,inplace=True)
    
    return dfObj

REDUCE SIZE OF DATAFRAME (**reduce_memory**)

Modifying the data types for values contained within select columns will drastically reduce the memory usage of the dataframe object.

In [10]:
def reduce_memory(dfObj):
    """
    This function changes the dtypes for specific columns in 
    the DataFrame object in order to reduce its memory usage
    
    Parameter(s)
    ------------
    dfObj: Pandas DataFrame object
        The object to be modified by the function
        
    Returns
    -------
    dfObj: Pandas DataFrame object
    """
    # Truncate name for borough label: 'RICHMOND / STATEN ISLAND'
    dfObj['borough'] = dfObj.borough.replace('RICHMOND / STATEN ISLAND',
                                             'STATEN ISLAND')
    
    # Create list of all columns that contain ISO8601 datetime
    list_of_datetime_cols = [name for name in list(dfObj.columns) 
                             if 'datetime' in str(name)]

    # Convert dtypes for each element in list to datetime
    for name in list_of_datetime_cols:
        dfObj[name] = pd.to_datetime(dfObj[name],errors='coerce')
       
    # Create list of all columns that contain time duration
    list_of_numeric_cols = [name for name in list(dfObj.columns) 
                            if (('seconds' in str(name))|
                                ('severity' in str(name))|
                                ('cad' in str(name)))]

    # Convert dtypes for each element in list to numeric
    for name in list_of_numeric_cols:
        dfObj[name] = pd.to_numeric(dfObj[name],errors='coerce')
        
    # Convert columns to category dtypes to reduce size of dataframe object
    dfObj['borough'] = dfObj.borough.astype('category')
    dfObj['zipcode'] = dfObj.zipcode.astype('category')
    dfObj['held_indicator'] = dfObj.held_indicator.astype('category')
    dfObj['valid_dispatch_rspns_time_indc'] = dfObj.valid_dispatch_rspns_time_indc.astype('category')
    dfObj['valid_incident_rspns_time_indc'] = dfObj.valid_incident_rspns_time_indc.astype('category')
    dfObj['incident_dispatch_area'] = dfObj.incident_dispatch_area.astype('category')
    dfObj['incident_disposition_code'] = dfObj.incident_disposition_code.astype('category')
    
    return dfObj

REDESIGN THE DATAFRAME (**format_df**)

Construct a boolean series that represents the target variable (fatality) using the corresponding values in 'incident_disposition_code'. Also, apply aesthetic changes to help improve the readability of the dataframe object.

In [11]:
def format_df(dfObj):
    """
    This function creates a Pandas series for the target variable (fatality), 
    two Pandas series for a MultiIndex, and re-orders the columns 
    
    Parameter(s)
    ------------
    dfObj: Pandas DataFrame object
        The object to be modified by the function
        
    Returns
    -------
    dfObj: Pandas DataFrame object
    """
    # Create a series for the target variable: fatality
    dfObj['fatality'] = np.logical_or(dfObj.incident_disposition_code.astype('int64') == 83,
                                      dfObj.incident_disposition_code.astype('int64') == 96)

    # Create separate columns for the year and month of the incident
    dfObj['incident_year'] = pd.DatetimeIndex(dfObj.incident_datetime).year
    dfObj['incident_month'] = pd.DatetimeIndex(dfObj.incident_datetime).month

    # Reorder dataframe columns
    col_order = ['incident_year','incident_month','cad_incident_id',
                 'incident_datetime','borough','zipcode',
                 'initial_call_type','initial_severity_level_code',
                 'final_call_type','final_severity_level_code',
                 'held_indicator','first_assignment_datetime',
                 'incident_dispatch_area','valid_dispatch_rspns_time_indc',
                 'dispatch_response_seconds_qy','first_activation_datetime',
                 'first_on_scene_datetime','incident_travel_tm_seconds_qy',
                 'valid_incident_rspns_time_indc','incident_response_seconds_qy',
                 'first_to_hosp_datetime','first_hosp_arrival_datetime',
                 'incident_close_datetime','incident_disposition_code','fatality']
    dfObj=dfObj[col_order]
    
    return dfObj

### Data Cleansing: Apply functions

In [12]:
# Apply drop_errors to dataframe object
df = drop_errors(df)

In [13]:
# Apply drop_immaterial to dataframe object
df = drop_immaterial(df)

In [14]:
# Apply reduce_memory to dataframe object
df = reduce_memory(df)

In [15]:
# Apply format_df to dataframe object
df = format_df(df)

In [16]:
# Create a MultiIndex on 'incident_year', 'incident_month' and 'cad_incident_id'
df.set_index(['incident_year','incident_month','cad_incident_id'],inplace=True)

### Inspect clean dataframe

In [17]:
df.shape

(8156954, 22)

In [18]:
df.head()

incident_datetime borough  \
incident_year incident_month cad_incident_id                               
2013          1              130010001       2013-01-01 00:00:04   BRONX   
                             130010002       2013-01-01 00:00:19   BRONX   
                             130010004       2013-01-01 00:01:04  QUEENS   
                             130010005       2013-01-01 00:01:16   BRONX   
                             130010006       2013-01-01 00:01:26   BRONX   

                                              zipcode initial_call_type  \
incident_year incident_month cad_incident_id                              
2013          1              130010001        10472.0            RESPIR   
                             130010002        10454.0              CARD   
                             130010004        11418.0            ARREST   
                             130010005        10453.0              SICK   
                             130010006        10457.0            INJURY   

                                              initial_severity_level_code  \
incident_year incident_month cad_incident_id                                
2013          1              130010001                                  4   
                             130010002                                  3   
                             130010004                                  1   
                             130010005                                  6   
                             130010006                                  5   

                                             final_call_type  \
incident_year incident_month cad_incident_id                   
2013          1              130010001                RESPIR   
                             130010002                  CARD   
                             130010004                ARREST   
                             130010005                  SICK   
                             130010006                INJURY   

                                              final_severity_level_code  \
incident_year incident_month cad_incident_id                              
2013          1              130010001                                4   
                             130010002                                3   
                             130010004                                1   
                             130010005                                6   
                             130010006                                5   

                                             held_indicator  \
incident_year incident_month cad_incident_id                  
2013          1              130010001                    N   
                             130010002                    N   
                             130010004                    N   
                             130010005                    N   
                             130010006                    N   

                                             first_assignment_datetime  \
incident_year incident_month cad_incident_id                             
2013          1              130010001             2013-01-01 00:01:45   
                             130010002             2013-01-01 00:01:18   
                             130010004             2013-01-01 00:01:33   
                             130010005             2013-01-01 00:02:12   
                             130010006             2013-01-01 00:01:58   

                                             incident_dispatch_area   ...     \
incident_year incident_month cad_incident_id                          ...      
2013          1              130010001                           B3   ...      
                             130010002                           B1   ...      
                             130010004                           Q3   ...      
                             130010005                           B2   ...      
                             130

In [19]:
df.info(verbose=True,null_counts=True,memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 8156954 entries, (2013, 1, 130010001) to (2018, 12, 183654386)
Data columns (total 22 columns):
incident_datetime                 8156954 non-null datetime64[ns]
borough                           8156954 non-null category
zipcode                           8011600 non-null category
initial_call_type                 8156954 non-null object
initial_severity_level_code       8156954 non-null int64
final_call_type                   8156954 non-null object
final_severity_level_code         8156954 non-null int64
held_indicator                    8156954 non-null category
first_assignment_datetime         8156954 non-null datetime64[ns]
incident_dispatch_area            8156954 non-null category
valid_dispatch_rspns_time_indc    8156954 non-null category
dispatch_response_seconds_qy      8156954 non-null int64
first_activation_datetime         8148356 non-null datetime64[ns]
first_on_scene_datetime           8156954 non-null datetime64[ns]
inc

In [20]:
df.memory_usage(deep=True)

Index                             114197779
incident_datetime                  65255632
borough                             8157504
zipcode                            16326044
initial_call_type                 505104602
initial_severity_level_code        65255632
final_call_type                   505201811
final_severity_level_code          65255632
held_indicator                      8157158
first_assignment_datetime          65255632
incident_dispatch_area              8160358
valid_dispatch_rspns_time_indc      8157096
dispatch_response_seconds_qy       65255632
first_activation_datetime          65255632
first_on_scene_datetime            65255632
incident_travel_tm_seconds_qy      65255632
valid_incident_rspns_time_indc      8157096
incident_response_seconds_qy       65255632
first_to_hosp_datetime             65255632
first_hosp_arrival_datetime        65255632
incident_close_datetime            65255632
incident_disposition_code           8157354
fatality                        

### Export dataframe to CSV

In [21]:
# Export dataframe to CSV
output_path = '../data/clean_EMS_data_from_csv.csv'
print('Exporting dataframe to CSV...')
df.to_csv(output_path,index=False,compression='gzip')
print('Dataframe successfully exported to CSV using \'gzip\' compression.')

Exporting dataframe to CSV...
Dataframe successfully exported to CSV using 'gzip' compression.


###### Summary
The original dataset was comprised of 8,557,848 observations of mixed data types, and its source file occupied approximately 1.98 GB of hard disk space and 10.9 GB in system memory. The target variable ("fatality") was created by applying a boolean filter on the "incident_disposition_code" column within the dataframe, which indicates the outcome of any EMS incident. 

After all data pre-processing was complete, the resulting clean dataset was comprised of 8,156,954 observations of mixed data types, with a clear target variable and 21 predictor variables. Its output file occupied 294 MB of hard disk space and 1.8 GB in system memory.